opt: avoid delete cascade fast path for regional by row tables

DrewKimball opened the pull request:
#### opt,testcat: allow defining the region column explicitly in opt tests

This commit adds support for explicitly defining the `crdb_region` column
for REGIONAL BY ROW tables in optimizer tests. This will allow testing
scenarios like a computed region column.

Epic: None

Release note: None

#### opt: avoid delete cascade fast path for regional by row tables

We have a fast path for cascading deletes where the filters from the
parent table are applied directly to the child table instead of joining
from a buffer to the child table. This can result in an efficient
`DeleteRange` operation in some cases, but in others, can result in
a "less constrained" scan on the child table. In particular, it can
result in visiting more regions than necessary for a regional-by-row
child table.

This commit prevents the fast path from applying for regional-by-row
child tables unless the region column is restricted to a single value.
It is possible to revert to the old behavior by disabling a new session
var: `optimizer_disable_cross_region_cascade_fast_path_for_rbr_tables`.

Informs: #146705

Release note (performance improvement): The optimizer will no longer
apply a fast-path to deletes cascading to regional-by-row tables. This
prevents the cascading delete from accessing more regions than necessary.

DrewKimball committed:
opt: avoid delete cascade fast path for regional by row tables

We have a fast path for cascading deletes where the filters from the
parent table are applied directly to the child table instead of joining
from a buffer to the child table. This can result in an efficient
`DeleteRange` operation in some cases, but in others, can result in
a "less constrained" scan on the child table. In particular, it can
result in visiting more regions than necessary for a regional-by-row
child table.

This commit prevents the fast path from applying for regional-by-row
child tables unless the region column is restricted to a single value.
It is possible to revert to the old behavior by disabling a new session
var: `optimizer_disable_cross_region_cascade_fast_path_for_rbr_tables`.

Informs: #146705

Release note (performance improvement): The optimizer will no longer
apply a fast-path to deletes cascading to regional-by-row tables. This
prevents the cascading delete from accessing more regions than necessary.

mgartner commented:
This problem affects non-MR tables too, as you've noted in #146705. Is it worth while to make this fix non-MR specific? It seems like we should never consider it if a prefix of index columns are not held constant by the filters.

Ex:

```
exec-ddl
CREATE TABLE parent (
  r STRING,
  k INT,
  v INT,
  PRIMARY KEY (r, k)
)
----

exec-ddl
CREATE TABLE child (
  r STRING,
  k INT,
  p_k INT,
  PRIMARY KEY (r, k),
  FOREIGN KEY (r, k) REFERENCES parent(r, k) ON DELETE CASCADE
)
----

opt-post-queries
DELETE FROM parent WHERE k = 4
----
root
 ├── delete parent
 │    ├── columns: <none>
 │    ├── fetch columns: r:6 k:7
 │    ├── cascades
 │    │    └── child_r_k_fkey
 │    ├── cardinality: [0 - 0]
 │    ├── volatile, mutations
 │    └── select
 │         ├── columns: r:6!null k:7!null
 │         ├── key: (6)
 │         ├── fd: ()-->(7)
 │         ├── scan parent
 │         │    ├── columns: r:6!null k:7!null
 │         │    ├── flags: avoid-full-scan
 │         │    └── key: (6,7)
 │         └── filters
 │              └── k:7 = 4 [outer=(7), constraints=(/7: [/4 - /4]; tight), fd=()-->(7)]
 └── cascade
      └── delete child
           ├── columns: <none>
           ├── fetch columns: child.r:16 child.k:17
           ├── cardinality: [0 - 0]
           ├── volatile, mutations
           └── select
                ├── columns: child.r:16!null child.k:17!null
                ├── key: (16)
                ├── fd: ()-->(17)
                ├── scan child
                │    ├── columns: child.r:16!null child.k:17!null
                │    ├── flags: avoid-full-scan
                │    └── key: (16,17)
                └── filters
                     └── child.k:17 = 4 [outer=(17), constraints=(/17: [/4 - /4]; tight), fd=()-->(17)]
```





<!-- Sent from Reviewable.io -->

DrewKimball commented:
>This problem affects non-MR tables too, as you've noted in https://github.com/cockroachdb/cockroach/issues/146705. Is it worth while to make this fix non-MR specific? It seems like we should never consider it if a prefix of index columns are not held constant by the filters.

I think the ideal end state will be that we do this as an exploration rule instead of during opt-building. One issue I see with only applying for an index prefix is that we may be able to infer other filters from check constraints or something, in which case making this change more general could cause a perf regression. I'd prefer to keep this fix specific to MR for now (and leave the issue open).

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I agree that this would ideally be an exploration rule. And great point about inferring filters (example of that below for posterity). I think it should be possible to determine if the prefix columns can be inferred from the filters—I think we already do something similar. But I understand that might be more work than we're willing to take on right now, so I'm with this approach for now.

IIUC we would expect this to cause a regression if the region column is a computed column based on the PK. Correct?


----

```
exec-ddl
CREATE TABLE parent (
  r STRING,
  k INT,
  v INT,
  PRIMARY KEY (r, k)
)
----

exec-ddl
CREATE TABLE child (
  r STRING AS (CASE WHEN k < 100 THEN 'a' ELSE 'b' END) STORED,
  k INT,
  p_k INT,
  PRIMARY KEY (r, k),
  FOREIGN KEY (r, k) REFERENCES parent(r, k) ON DELETE CASCADE
)
----

opt-post-queries
DELETE FROM parent WHERE k = 4
----
root
 ├── delete parent
 │    ├── columns: <none>
 │    ├── fetch columns: r:6 k:7
 │    ├── cascades
 │    │    └── child_r_k_fkey
 │    ├── cardinality: [0 - 0]
 │    ├── volatile, mutations
 │    └── select
 │         ├── columns: r:6!null k:7!null
 │         ├── key: (6)
 │         ├── fd: ()-->(7)
 │         ├── scan parent
 │         │    ├── columns: r:6!null k:7!null
 │         │    ├── flags: avoid-full-scan
 │         │    └── key: (6,7)
 │         └── filters
 │              └── k:7 = 4 [outer=(7), constraints=(/7: [/4 - /4]; tight), fd=()-->(7)]
 └── cascade
      └── delete child
           ├── columns: <none>
           ├── fetch columns: child.r:16 child.k:17
           ├── cardinality: [0 - 0]
           ├── volatile, mutations
           └── scan child
                ├── columns: child.r:16!null child.k:17!null
                ├── constraint: /16/17: [/'a'/4 - /'a'/4]
                ├── flags: avoid-full-scan
                ├── cardinality: [0 - 1]
                ├── key: ()
                └── fd: ()-->(16,17)
```

> Reviewed 10 of 10 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148105#-:-OSzqdDHES4VRJ6KdYUd:b-clki8v)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rytaft)

___
*[`pkg/sql/opt/optbuilder/fk_cascade.go` line 248 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148105#-OSzsK7w9D7Rj7u3ffm7:-OSzsK7x7AXWS9QquASP:b3zngie) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8dc80d6eca69eb364fdfbdaca8d92829c96e638/pkg/sql/opt/optbuilder/fk_cascade.go#L248)):*
> ```Go
> 				if eq, isEq := filters[i].Condition.(*memo.EqExpr); isEq {
> 					if v, leftIsVar := eq.Left.(*memo.VariableExpr); leftIsVar && v.Col == regionalColID {
> 						if _, rightIsConst := eq.Right.(*memo.ConstExpr); rightIsConst {
> ```

nit: you can use `opt.IsConstValueOp` instead of a type assertion
___
*[`pkg/sql/vars.go` line 4039 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148105#-OSzsxq1Er3jckEg5Tc3:-OSzsxq1Er3jckEg5Tc4:b-9juo9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8dc80d6eca69eb364fdfbdaca8d92829c96e638/pkg/sql/vars.go#L4039)):*
> ```Go
> 
> 	// CockroachDB extension.
> 	`optimizer_always_use_cascade_fast_path_for_rbr_tables`: {
> ```

nit: I find "always" in the session name confusing because if there a number of requirements to use the cascade fast path even if this is set to true. Should we invert it and name it `optimizer_disable_cross_region_cascade_fast_path_for_rbr_tables`? Not sure that's much better I guess...


<!-- Sent from Reviewable.io -->

DrewKimball committed:
opt,testcat: allow defining the region column explicitly in opt tests

This commit adds support for explicitly defining the `crdb_region` column
for REGIONAL BY ROW tables in optimizer tests. This will allow testing
scenarios like a computed region column.

Epic: None

Release note: None

DrewKimball reviewed (commented):
>IIUC we would expect this to cause a regression if the region column is a computed column based on the PK. Correct?

Ugh, that's a good point that I hadn't considered. It would be a regression if all referenced columns were constrained to be constant in the filters + check constraints. I'll add a test case - I think it makes sense to track improving that in #146705 (which won't close after this merges).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148105#-:-OT-J3N38x0i6xisFcNQ:b1ft6o4)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @rytaft)

___
*[`pkg/sql/vars.go` line 4039 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148105#-OSzsxq1Er3jckEg5Tc3:-OT05NC-2ni_utqsoeZV:bakzp34) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8dc80d6eca69eb364fdfbdaca8d92829c96e638/pkg/sql/vars.go#L4039)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: I find "always" in the session name confusing because if there a number of requirements to use the cascade fast path even if this is set to true. Should we invert it and name it `optimizer_disable_cross_region_cascade_fast_path_for_rbr_tables`? Not sure that's much better I guess...
</blockquote></details>

I changed it to `optimizer_disable_cross_region_cascade_fast_path_for_rbr_tables` - bit of a mouthful, but it's not meant to be used often anyway :)
___
*[`pkg/sql/opt/optbuilder/fk_cascade.go` line 248 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148105#-OSzsK7w9D7Rj7u3ffm7:-OT-Iv7i9CLLN5xwHN3A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8dc80d6eca69eb364fdfbdaca8d92829c96e638/pkg/sql/opt/optbuilder/fk_cascade.go#L248)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: you can use `opt.IsConstValueOp` instead of a type assertion
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner commented:
> > IIUC we would expect this to cause a regression if the region column is a computed column based on the PK. Correct?
>
> Ugh, that's a good point that I hadn't considered. It would be a regression if all referenced columns were constrained to be constant in the filters + check constraints. I'll add a test case - I think it makes sense to track improving that in https://github.com/cockroachdb/cockroach/issues/146705 (which won't close after this merges).

I think it's a minor regression though. The cost of buffering in the test you added for this is minor compared to the cost of a cross-region round-trip in the other cases this PR is fixing, so I think it's an acceptable sacrifice.
